### PR TITLE
feat: css webpackIgnoreComment

### PR DESCRIPTION
**Pull request by faga295**
**What kind of change does this PR introduce?**

**Did you add tests for your changes?**

**Does this PR introduce a breaking change?**

**What needs to be documented once your changes are merged?**

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
Please add all tests from css-loader

**Comment by faga295** [comment]
Now image-set can't pass string type like `image-set("./img1x.png" 1x);`, so a test case in css-loader now can't pass

File: lib/css/walkCssTokens.js [hunk 1]
```diff
@@ -733,16 +747,25 @@ module.exports.eatWhitespace = (input, pos) => {
...
+ * @param {number} direction consume direction -1: reverse, 1: forward
  * @returns {number} position after whitespace and comments
  */
-module.exports.eatWhitespaceAndComments = (input, pos) => {
+module.exports.eatWhitespaceAndComments = (
+	input,
+	pos,
+	callbacks,
+	direction = 1
+) => {
```
Comment (alexander-akait) [inline lib/css/walkCssTokens.js:763]: I think this is not a correct implementation, our parser is written according to the specification and we should almost not touch it, except adding callbacks

File: lib/css/walkCssTokens.js [hunk 2]
```diff
@@ -140,18 +147,23 @@ const consumeComments = (input, pos, _callbacks) => {
...
 	) {
-		pos += 1;
-		while (pos < input.length) {
+		const start = pos;
+		pos += direction;
+		while (direction > 0 ? pos < input.length : pos > -1) {
 			if (
 				input.charCodeAt(pos) === CC_ASTERISK &&
-				input.charCodeAt(pos + 1) === CC_SOLIDUS
+				input.charCodeAt(pos + direction) === CC_SOLIDUS
```
Comment (alexander-akait) [inline lib/css/walkCssTokens.js:157]: It should be not changed

File: lib/css/CssParser.js [hunk 3]
```diff
@@ -829,6 +980,7 @@ class CssParser extends Parser {
 								return processLocalDeclaration(input, start, end);
 							}
 						}
+						if (!currentDeclaration) processDeclaration(input, start, end);
```
Comment (alexander-akait) [inline lib/css/CssParser.js:983]: Wrong

Comment (alexander-akait) [reply lib/css/CssParser.js:983]: You don't need at all extra processing, just use binary search before creation deps

**Review by alexander-akait — CHANGES_REQUESTED** [review]
This code should be simplified, you don't need extra processing, just add a comment callback, collect all comments and use binary search before creation of deps

File: lib/css/CssParser.js [hunk 4]
```diff
@@ -526,12 +610,51 @@ class CssParser extends Parser {
...
+								start
+							]).options.needIgnore;
+							if (
+								stringNeedIgnore !== undefined
+									? stringNeedIgnore
+									: functionNeedIgnore !== undefined
+										? functionNeedIgnore
+										: lineNeedIgnore
+							)
+								break;
```
[the remark this answers is not in the document]
Comment (faga295) [reply lib/css/CssParser.js:657]:
```
.class {
  color: red;
  /** webpackIgnore: true */
  background: /** webpackIgnore: false */ url("./url/img.png")
}
```
`lineNeedIgnore` indicates last line webpackIgnore, `functionNeedIngore` indicates webpackIgnore before url(), if we don't collect this, I think we can parseComments from the end position of last declaration to the start position of url function, but this parse range is a little large, maybe could cause some bugs.

[the remark this answers is not in the document]
Comment (faga295) [reply lib/css/CssParser.js:657]: We should record some key positions instead of use `eatWhitespaceAndComments`

File: lib/css/CssParser.js [hunk 5]
```diff
@@ -212,14 +218,28 @@ class CssParser extends Parser {
...
+			comment: (input, start, end) => {
+				if (!this.comments) this.comments = [];
+				this.comments.push({
+					value: input.slice(start + 2, end - 2).trim(),
+					start,
+					end
+				});
+				return end;
+			}
+		};
```
Comment (alexander-akait) [inline lib/css/CssParser.js:235]: It should be in `Cssparser`

File: lib/css/CssParser.js [hunk 6]
```diff
@@ -303,7 +326,11 @@ class CssParser extends Parser {
 		 * @returns {number} position after parse
 		 */
 		const parseExports = (input, pos) => {
-			pos = walkCssTokens.eatWhitespaceAndComments(input, pos);
+			pos = walkCssTokens.eatWhitespaceAndComments(
+				input,
+				pos,
+				commentCallbacks
```
Comment (alexander-akait) [inline lib/css/CssParser.js:332]: Always pass `callbacks` everything

File: lib/css/CssParser.js [hunk 7]
```diff
@@ -1044,6 +1203,58 @@ class CssParser extends Parser {
...
+		const comments = /** @type {Comment[]} */ this.comments;
+		let idx = binarySearchBounds.ge(comments, rangeStart, compare);
+		const commentsInRange = [];
+		while (comments[idx] && comments[idx].end <= rangeEnd) {
+			commentsInRange.push(comments[idx]);
+			idx++;
+		}
+
+		return commentsInRange;
+	}
```
Comment (alexander-akait) [inline lib/css/CssParser.js:1228]: It should cssparser.js

File: lib/css/CssParser.js [hunk 8]
```diff
@@ -1044,6 +1203,58 @@ class CssParser extends Parser {
...
+				} catch (err) {
+					const newErr = new Error(String(err.message));
+					newErr.stack = String(err.stack);
+					Object.assign(newErr, { comment });
+					errors.push(newErr);
+				}
+			}
+		}
+		return { options, errors };
+	}
```
Comment (alexander-akait) [inline lib/css/CssParser.js:1257]: It should cssparser.js

File: lib/css/CssParser.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexander-akait) [reply lib/css/CssParser.js:657]: You have `locConverter` to calculate line before

**Review by alexander-akait — COMMENTED** [review]
I have said many times that `walkCssTokens.js` changes should be minimal or even be absent, it will always lead to problems, the current logic is written according to the specification, and should not change at all

File: lib/css/CssParser.js (hunk 5 above)
Comment (faga295) [reply lib/css/CssParser.js:235]: I beleive this is in cssParser.js

**Comment by faga295** [comment]
I still reserve mulitple ignores like `declarationIgnore`,`needIgnore` and `stringIgnore`, which I think are necessary to pass all the related test cases in css-loader.However, the wrong changes of `walkCssTokens` are removed, and I use another way to implement this

**Comment by alexander-akait** [comment]
Fixed

**Pull request closed without merging**